src/workerd/api/sockets.h

```diff
@@ -10,86 +10,28 @@
-
-  // AsyncInputStream methods:
-  kj::Promise<size_t> read(void* buffer, size_t minBytes, size_t maxBytes) override;
-  kj::Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override;
-
-  // AsyncOutputStream methods:
-  kj::Promise<void> write(const void* buffer, size_t size) override;
-  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override;
-
-  kj::Promise<void> whenWriteDisconnected() override;
-
-private:
-  template <typename T>
-  kj::Promise<T> thenOrRunNow(std::function<kj::Promise<T> (kj::Own<kj::AsyncIoStream> *)> f) {
-    // Either calls `then` on the `inner` promise with the `f` callback, or calls `f` on an already
-    // stored stream (if the `inner` promise completed).
-    KJ_IF_MAYBE(e, error) {
-      kj::throwRecoverableException(kj::mv(*e));
-    }
-
-    KJ_IF_MAYBE(s, completedInner) {
-      return f(s);
-    } else {
-      return inner.then([this, f](kj::Own<kj::AsyncIoStream> stream) -> kj::Promise<T> {
-        completedInner = kj::mv(stream);
-        return f(&KJ_ASSERT_NONNULL(completedInner));
-      });
-    }
-  }
-
-  kj::Maybe<kj::Own<kj::AsyncIoStream>> completedInner;
-  // Stored io stream once the `inner` promise completes.
-  kj::Maybe<kj::Exception> error;
-  // Stored error if any of the operations throw.
-  kj::Promise<kj::Own<kj::AsyncIoStream>> inner;
-  // A promise holding a stream that will be available at some point in the future. Typically once
-  // a connection is established.
-};
-
 struct SocketOptions {
   jsg::Unimplemented tls; // TODO(later): TCP socket options need to be implemented.
   JSG_STRUCT(tls);
 };
 
-struct InitData {
-  jsg::Ref<ReadableStream> readable;
-  jsg::Ref<WritableStream> writable;
-  IoOwn<kj::PromiseFulfillerPair<void>> closeFulfiller;
-};
-
 class Socket: public jsg::Object {
 public:
-  Socket(InitData data)
-      : readable(kj::mv(data.readable)), writable(kj::mv(data.writable)),
-        closeFulfiller(kj::mv(data.closeFulfiller)) {};
-  Socket(kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise);
+  Socket(jsg::Lock& js, jsg::Ref<ReadableStream> readable, jsg::Ref<WritableStream> writable,
+      kj::Own<jsg::PromiseResolverPair<void>> close)
+      : readable(kj::mv(readable)), writable(kj::mv(writable)),
+        closeFulfiller(kj::mv(close)),
+        closedPromise(kj::mv(closeFulfiller->promise)),
+        isClosed(false) {
+  };
 
   jsg::Ref<ReadableStream> getReadable() { return readable.addRef(); }
   jsg::Ref<WritableStream> getWritable() { return writable.addRef(); }
-  jsg::Promise<void> getClosed() {
-    // TODO: Right now this promise won't complete when the remote end closes the socket.
-    //       Do we need to wrap ReadableStream/WritableStream to make this work or is there a
-    //       better way?
-    auto& context = IoContext::current();
-    return context.awaitIo(kj::mv(closeFulfiller->promise));
+  jsg::MemoizedIdentity<jsg::Promise<void>>& getClosed() {
```

Comment:
It looks like, at present, this promise is only resolved when the application calls `close()` explicitly -- not when the remote side disconnects. Is that correct?

If so, it seems like it's not accomplishing a whole lot to offer a way to listen for closure. The application can set that up itself and save us the trouble.

If the intent is that this promise should also tell the application when the remote end disconnects the socket, then I think you should add code that listens on `whenWriteDisconnected()` and fulfills the promise at that point. (Or maybe it should wait on both `whenWriteDisconnected()` and EOF or error on the read side, and only resolve when both are true... I'm not really sure... I'd be satisfied with waiting for `whenWriteDisconnected()` only.)

Reply:
Yep, it's the current behaviour in workerd HEAD as well. I split this PR off from https://github.com/cloudflare/workerd/pull/205 to not block on this and instead focus on the refactoring.

I will follow up on this but first we need to decide how we want the semantics to work exactly.